presto-product-tests/src/main/java/com/facebook/presto/tests/SqlCancelTests.java

```diff
@@ -0,0 +1,197 @@
+
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static com.facebook.presto.tests.TestGroups.CANCEL_QUERY;
+import static com.google.common.util.concurrent.Uninterruptibles.sleepUninterruptibly;
+import static com.teradata.tempto.assertions.QueryAssert.assertThat;
+import static com.teradata.tempto.query.QueryExecutor.query;
+import static java.lang.String.format;
+import static java.util.concurrent.Executors.newFixedThreadPool;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.fail;
+
+public class SqlCancelTests
+        extends ProductTest
+{
+    private static final String TABLE_NAME = "cancel_query_table_name";
+    private static final long WAIT_TIME = 10000;
+    private static final long STEP = 100;
+
+    private ExecutorService executor;
+    private QueryCanceller queryCanceller;
+
+    @Inject
+    @Named("databases.presto.server_address")
+    private String serverAddress;
+
+    @BeforeTestWithContext
+    public void setUp()
+            throws Exception
+    {
+        executor = newFixedThreadPool(2);
+        queryCanceller = new QueryCanceller();
+        query(format("DROP TABLE IF EXISTS %s", TABLE_NAME));
+    }
+
+    @AfterTestWithContext
+    public void cleanUp()
+            throws Exception
+    {
+        if (executor != null) {
+            executor.shutdownNow();
+        }
+        if (queryCanceller != null) {
+            queryCanceller.close();
+        }
+        query(format("DROP TABLE IF EXISTS %s", TABLE_NAME));
+    }
+
+    private class SqlExecuteCallable
+            implements Callable<Boolean>
+    {
+        private String sql;
+
+        public SqlExecuteCallable(String sql)
+        {
+            this.sql = sql;
+        }
+
+        public Boolean call()
+        {
+            try {
+                query(sql);
+                return false;
+            }
+            catch (QueryExecutionException e) {
+                return true;
+            }
+        }
+    }
+
+    private class SqlCancelCallable
+            implements Callable<Response>
+    {
+        private String sqlStatement;
+
+        public SqlCancelCallable(String sqlStatement)
+        {
+            this.sqlStatement = sqlStatement;
+        }
+
+        public Response call()
+        {
+            long endTime = System.currentTimeMillis() + WAIT_TIME;
+            while (System.currentTimeMillis() < endTime && !Thread.currentThread().isInterrupted()) {
+                List<Object> queryResult = query(format("SELECT state, query_id from system.runtime.queries" +
+                        " WHERE query = '%s' ORDER BY created DESC LIMIT 1", sqlStatement)).row(0);
+                if (queryResult == null || queryResult.size() != 2) {
```

Comment:
This null check won't help.  You'll get an IndexOutOfBoundsException if there are no rows.  Check instead that getRowsCount > 0 before calling `row(0)`